crates/apollo-compiler/src/validation/selection.rs

```diff
@@ -188,12 +196,9 @@ pub(crate) fn same_response_shape(
             ));
             let grouped_by_name = group_fields_by_name(merged_set);
 
-            for (_, fields_for_name) in grouped_by_name {
+            for fields_for_name in grouped_by_name.values() {
                 // 9. Given each pair of members subfieldA and subfieldB in fieldsForName:
-                let Some((subfield_a, rest)) = fields_for_name.split_first() else {
-                    continue;
-                };
-                for subfield_b in rest {
+                for (subfield_a, subfield_b) in pair_combinations(fields_for_name) {
```

Comment:
Do we expect real-world cases where `fields_for_name` can grow very large? Checking all pairs this way takes quadratic time. Trying to build a merged selection set just to throw it away seems wasteful but I think could be done in linear time.

Reply:
This kinda sounds like https://tech.new-work.se/graphql-overlapping-fields-can-be-merged-fast-ea6e92e0a01 which i still have as TODO to read. So maybe for later

Reply:
That was my thinking, but now comparing some crafted inputs to graphql-js, it looks like the JS implementation has some changes to speed it up dramatically while my implementation falls over. I had assumed that graphql-js was using the naive implementation. I'll see about implementing either the other algorithm or the graphql-js changes tomorrow, so we don't introduce a new easy DOS vector